cmd/opampsupervisor/supervisor/supervisor.go

```diff
@@ -0,0 +1,649 @@
+		effectiveConfigFilePath: "effective.yaml",
+	}
+
+	if err := s.loadConfig(configFile); err != nil {
+		return nil, fmt.Errorf("error loading config: %w", err)
+	}
+
+	s.createInstanceID()
+	logger.Debugf("Supervisor starting, id=%v, type=%s, version=%s.",
+		s.instanceID.String(), agentType, agentVersion)
+
+	s.loadAgentEffectiveConfig()
+
+	if err := s.startOpAMP(); err != nil {
+		return nil, fmt.Errorf("cannot start OpAMP client: %w", err)
+	}
+
+	var err error
+	s.commander, err = commander.NewCommander(
+		s.logger,
+		s.config.Agent,
+		"--config", s.effectiveConfigFilePath,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	s.startHealthCheckTicker()
+	go s.runAgentProcess()
+
+	return s, nil
+}
+
+func (s *Supervisor) loadConfig(configFile string) error {
+	if configFile == "" {
+		return errors.New("path to config file cannot be empty")
+	}
+
+	k := koanf.New("::")
+	if err := k.Load(file.Provider(configFile), yaml.Parser()); err != nil {
+		return err
+	}
+
+	if err := k.Unmarshal("", &s.config); err != nil {
+		return fmt.Errorf("cannot parse %v: %w", configFile, err)
+	}
+
+	return nil
+}
+
+func (s *Supervisor) startOpAMP() error {
+	s.opampClient = client.NewWebSocket(s.logger)
+
+	settings := types.StartSettings{
+		OpAMPServerURL: s.config.Server.Endpoint,
+		InstanceUid:    s.instanceID.String(),
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc: func() {
+				s.logger.Debugf("Connected to the server.")
+			},
+			OnConnectFailedFunc: func(err error) {
+				s.logger.Errorf("Failed to connect to the server: %v", err)
+			},
+			OnErrorFunc: func(err *protobufs.ServerErrorResponse) {
+				s.logger.Errorf("Server returned an error response: %v", err.ErrorMessage)
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				return s.createEffectiveConfigMsg(), nil
+			},
+			OnMessageFunc: s.onMessage,
+		},
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_AcceptsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsEffectiveConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsOwnMetrics |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsHealth,
+	}
+	err := s.opampClient.SetAgentDescription(s.createAgentDescription())
+	if err != nil {
+		return err
+	}
+
+	err = s.opampClient.SetHealth(&protobufs.AgentHealth{Healthy: false})
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debugf("Starting OpAMP client...")
+
+	err = s.opampClient.Start(context.Background(), settings)
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debugf("OpAMP Client started.")
+
+	return nil
+}
+
+func (s *Supervisor) createInstanceID() {
+	// Generate instance id.
+	entropy := ulid.Monotonic(rand.New(rand.NewSource(0)), 0)
+	s.instanceID = ulid.MustNew(ulid.Timestamp(time.Now()), entropy)
+
+	// TODO: set instanceId in the Collector config.
```

Comment:
The Supervisor also needs to persist the instance id so that it does not change on every Supervisor restart.

See https://docs.google.com/document/d/1KtH5atZQUs9Achbce6LiOaJxLbksNJenvgvyKLsJrkc/edit#heading=h.8szgxmjard0e